### ISSUE TITLE
Automatic arena detection

### ISSUE THREAD

**Issue by 2909ft** [mlfpm/deepof#24]
Hello!

Thank you for creating DeepOF! I am currently exploring it, and it seems quite promising. I have one suggestion so far: would it be possible to make the arena detection (for the polygon scenario) a bit more efficient? Either through automatic detection, or at least through copying the polygon from the first video to all the others (the arena is always in the same place in my case). I have a lot of videos and it is very time consuming to draw the arena in all of them.

Cheers,
Filipa

**Comment by lucasmiranda42** [comment]
Hi, Filipa! And thank you for your suggestion, and overall interest in DeepOF!

This is something we're looking into at the moment, although there is no clear decision so far on how to proceed. The most likely scenario will involve an automatic segmentation model (such as [SAM](https://github.com/facebookresearch/segment-anything)), which will still require user input, but should be way easier to use. Copying the selected polygon to all videos should also be a relatively easy feature to implement. This should be a part of the next release :)

If you have a clear idea in mind, feel free to submit a pull request! (see our [contributions guide](https://github.com/mlfpm/deepof/blob/master/CONTRIBUTING.md) in the readme).

For now, I'll leave the issue open and keep you posted when the new version is up!

Best,
Lucas

**Comment by lucasmiranda42** [comment]
Dear @2909ft,

I'm happy to tell you that we have finally integrated improvements to the arena detection pipeline to our [development branch](https://gitlab.mpcdf.mpg.de/lucasmir/deepof/-/tree/development). While you wait for full integration into the main branch and an update to the documentation, here is a simple overview of how it works:

For starters, the `arena` parameter in the `Project` constructor now takes four different values: `circular-autodetect`, `polygonal-autodetect`, `circular-manual`, and `polygonal-manual`. The first two are recommended when creating the projects, and respectively cover situations in which you have circular (or elliptical) and polygonal arenas.

When running `.create()` with `circular-autodetect`, DeepOF will take care of everything for you (no need to tag anything manually!). When selecting `polygonal-autodetect` (as I believe is your case), you will now be prompted with manual arena selection **just once**. DeepOF will then use the selected arena as a template prompt for [SAM](https://segment-anything.com/), which will automatically detect all arenas. Moreover, a new folder will appear in your project directory, labelled as `Arena_detection`. Inside this folder, you'll see samples of the detected arena for each video in your dataset. In case any detection failed, you can still run `.edit_arenas()` (with `circular-manual` or `polygonal-manual`) as specified in the documentation. This should massively reduce the workload for large datasets, as you should only manually label those edge cases where the initial segmentation did not yield a satisfactory result.

Last but not least, remember that, when using polygonal arenas, the first edge you select in the GUI should match the length in millimetres specified in the `video_scale` parameter in the `Project` constructor. If you open any of the detected arena samples, you'll see two large circles on the corners of the arena that DeepOF detects should match the edge you selected:

<img width="628" alt="image" src="https://github.com/mlfpm/deepof/assets/45250053/27c03965-cd8b-43e8-b17a-3794832f6d15">


Best, and feel free to open a new issue if you find any problems or need assistance!
Lucas

**Comment by 2909ft** [comment]
Thank you for your reply! Very nice!

Is it possible to explore the development branch in the meantime? By git cloning it?

**Comment by lucasmiranda42** [comment]
Yes, of course! You can just run `git clone https://gitlab.mpcdf.mpg.de/lucasmir/deepof.git`, and, once within the cloned repo, run `git switch development`. Let me know if you don't manage!
